File: scripts/generate-icons.js
const fs = require('fs');
const path = require('path');

const iconDir = path.join(__dirname, '../images/tab');

const createPNG = (width, height, r, g, b, a = 255) => {
  const signature = Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]);
  
  const createChunk = (type, data) => {
    const length = Buffer.alloc(4);
    length.writeUInt32BE(data.length);
    const typeBuffer = Buffer.from(type);
    const crcData = Buffer.concat([typeBuffer, data]);
    let crc = 0xFFFFFFFF;
    for (let i = 0; i < crcData.length; i++) {
      crc ^= crcData[i];
      for (let j = 0; j < 8; j++) {
        crc = (crc >>> 1) ^ (crc & 1 ? 0xEDB88320 : 0);
      }
    }
    crc = (crc ^ 0xFFFFFFFF) >>> 0;
    const crcBuffer = Buffer.alloc(4);
    crcBuffer.writeUInt32BE(crc);
    return Buffer.concat([length, typeBuffer, data, crcBuffer]);
  };
  
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = 6;
  ihdr[10] = 0;
  ihdr[11] = 0;
  ihdr[12] = 0;
  
  const rawData = [];
  for (let y = 0; y < height; y++) {
    rawData.push(0);
    for (let x = 0; x < width; x++) {
      rawData.push(r, g, b, a);
    }
  }
  
  const zlib = require('zlib');
  const compressed = zlib.deflateSync(Buffer.from(rawData));
  
  const ihdrChunk = createChunk('IHDR', ihdr);
  const idatChunk = createChunk('IDAT', compressed);
  const iendChunk = createChunk('IEND', Buffer.alloc(0));
  
  return Buffer.concat([signature, ihdrChunk, idatChunk, iendChunk]);
};

const icons = {
  'calc': { normal: [153, 153, 153], active: [255, 107, 0] },
  'calc_active': { normal: [255, 107, 0], active: [255, 107, 0] },
  'rebate': { normal: [153, 153, 153], active: [255, 107, 0] },
  'rebate_active': { normal: [255, 107, 0], active: [255, 107, 0] },
  'index': { normal: [153, 153, 153], active: [255, 107, 0] },
  'index_active': { normal: [255, 107, 0], active: [255, 107, 0] },
  'store': { normal: [153, 153, 153], active: [255, 107, 0] },
  'store_active': { normal: [255, 107, 0], active: [255, 107, 0] }
};

if (!fs.existsSync(iconDir)) {
  fs.mkdirSync(iconDir, { recursive: true });
}

for (const [name, colors] of Object.entries(icons)) {
  const [r, g, b] = colors.normal;
  const png = createPNG(48, 48, r, g, b);
  fs.writeFileSync(path.join(iconDir, `${name}.png`), png);
  console.log(`Created ${name}.png`);
}

console.log('All icons generated!');

File: scripts/generate-icons.py
import struct
import zlib
import os

def create_simple_png(width, height, r, g, b, a=255):
    signature = b'\x89PNG\r\n\x1a\n'
    
    def chunk(chunk_type, data):
        c = chunk_type + data
        crc = zlib.crc32(c) & 0xffffffff
        return struct.pack('>I', len(data)) + c + struct.pack('>I', crc)
    
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0)
    
    raw_data = bytearray()
    for y in range(height):
        raw_data.append(0)
        for x in range(width):
            raw_data.extend([r, g, b, a])
    
    compressed = zlib.compress(bytes(raw_data))
    
    return (signature + 
            chunk(b'IHDR', ihdr_data) + 
            chunk(b'IDAT', compressed) + 
            chunk(b'IEND', b''))

def is_valid_png(data):
    if len(data) < 8:
        return False
    if data[:8] != b'\x89PNG\r\n\x1a\n':
        return False
    return True

icon_dir = os.path.join(os.path.dirname(__file__), '../images/tab')

GRAY = (153, 153, 153)
ORANGE = (255, 107, 0)

icons = {
    'coupon.png': GRAY,
    'coupon_active.png': ORANGE,
    'exchange.png': GRAY,
    'exchange_active.png': ORANGE,
    'compare.png': GRAY,
    'compare_active.png': ORANGE,
    'buyer.png': GRAY,
    'buyer_active.png': ORANGE,
    'profile.png': GRAY,
    'profile_active.png': ORANGE
}

os.makedirs(icon_dir, exist_ok=True)

for name, (r, g, b) in icons.items():
    filepath = os.path.join(icon_dir, name)
    png_data = create_simple_png(48, 48, r, g, b)
    
    if is_valid_png(png_data):
        with open(filepath, 'wb') as f:
            f.write(png_data)
        print(f'Created valid PNG: {name} ({len(png_data)} bytes)')
    else:
        print(f'ERROR: Invalid PNG: {name}')

print('\nAll icons generated and verified!')
